CAMEL-10885 Add mask option to log EIP

igarashitm opened the pull request:
Introduced a StringFormatter interface to plugin a formatter into the Log EIP processor. MaskingStringFormatter is the implementation which masks sensitive information like password and passphrase for key=value, XML and JSON.

Like Log Component looks for a ExchangeFormatter with the name "logFormatter" in registry to be used for a default formatter, LogDefinition looks for a "logEipFormatter" for the default formatter. It can be overridden by specifying formatterRef option in Log DSL. It also introduced a series of logWithFormatter()/logWithFormatterRef() to be used in Java DSL.

Any feedback would be really appreciated. As those addition of the methods doesn't look really cool, that would be great if you could have better idea...

In the meantime, as it seems there's no log-eip.adoc in the src/main/docs/eips yet, I'll look into how it could be generated and add more commit here once it succeeds.

davsclaus commented:
We have to be careful to not add too many methods in the Java DSL in the processor definition, eg all those new logWithFormatter... 

The log already have many methods in the Java DSL. The formatter is a bit overkill as people was looking for a way to avoid logging sensitive information, hence the mask=true|false. 

That we use some StringFormatter with regexp to search/replace as an implementation detail is fine. But I would like to keep the DSL simpler, and just have a boolean true|false for now.

We may also consider turning this on globally, maybe on <camelContext> where we can do a various things today to logMessageHistory=true|false etc. So we could have a logEipMask=true|false.

And then users should also be able to configure this in camel-spring-boot as well where you can turn this on|off globally on the context.

igarashitm commented:
Thanks for the comment @davsclaus !

Yes that's the point I'm still wondering - do we need mask=true|false flag at DSL level? it would still need to add an option to the log() DSL methods (or add logWithMask()). My wild guess is that in 99% of usecases the global flag at context level would be enough.

davsclaus commented:
Yeah lets start with the global flag first.

igarashitm commented:
Added logEipMask boolean option on both of CamelContext and RouteDefinition. I'll add log-eip.adoc later.

igarashitm commented:
Added adoc and exposed the logEipMask flag via JMX property and command.

davsclaus commented:
Thanks can you add this to camel-spring-boot as well
https://github.com/apache/camel/blob/master/components/camel-spring-boot/src/main/java/org/apache/camel/spring/boot/CamelConfigurationProperties.java#L259

davsclaus commented:
Also I wonder if we should add this to the log component? As it also logs the message body, and you may want to mask that as well? 
https://github.com/apache/camel/tree/master/camel-core/src/main/java/org/apache/camel/component/log

And if so maybe the option should be renamed from logEipMask to logMask ? or logMask

igarashitm commented:
Ah yes, I thought the log component was a next target, but having a common global flag it would be better to make it at the same time. Do we also mask headers and properties? My plan is to let DefaultExchangeFormatter use MaskingStringFormatter for body mask, and with same keywords it can also masks headers and properties in DefaultExchangeFormatter.

davsclaus commented:
Maybe for the log component we mask the entire string that is being logged, then it includes all headers / properties etc. 

I think at the end of the default exchange formatter you can do that, or maybe do that separately in the log component after the default exchange formatter, so they are separated.

igarashitm commented:
OK, property is ConcurrentHashMap and header is CaseInsensitiveMap, both of them are printed in key=value format. That's easier. I'll make that change!

igarashitm commented:
Added masking support for log component as well, with renaming logEipMask flag to logMask. Added some masking test for log component and updated both adoc.

davsclaus commented on camel-core/src/main/java/org/apache/camel/model/RouteDefinition.java:
```diff
@@ -1132,6 +1169,17 @@ protected RouteContext addRoutes(CamelContext camelContext, Collection<Route> ro
+            if (isLogMask != null) {
+                routeContext.setLogMask(isLogMask);
+                if (isLogMask) {
+                    log.debug("Security mask for Logging is enabled on route: {}", getId());
```
This option is not per route, but per CamelContext

igarashitm replied:
Should it be only context level? I added this as all other options in [RuntimeConfiguration](https://github.com/apache/camel/pull/1561/files#diff-936d95845082ad400b2c91c6b14e17c4) can be overridden at route level. It's only effective for log EIP though. Do you think we also remove the logMask flag from log component option as well so that the flag could be set only at context level?

davsclaus replied:
Ah sorry my bad about the RuntimeConfiguration being two levels. So lets keep this as-is. I will get this merged now. Then you can work on the spring boot support next.

davsclaus commented:
There was that one comment as posted above.

Also I did not see this added to camel-spring-boot, where you should be able to turn this on|off as well.

igarashitm commented:
Thanks for your review again! for spring-boot I added following two but is not enough? maybe I need to look into camel-spring-boot more.
https://github.com/apache/camel/pull/1561/files#diff-df49e68de94827df4cc89005af518f47
https://github.com/apache/camel/pull/1561/files#diff-3c917dd12d0fe81bf42f6b42ccbe848f

davsclaus commented:
Ah great it was there, its harder to spot when there is 30+ changed files in a big PR.

davsclaus commented:
Thanks for the PR it has been merged. Do you mind closing this?

igarashitm commented:
Yeah thank you so much for your patience on my messed up PR ;) closing.
